File: ServerCore/src/main/java/org/elkoserver/foundation/server/BaseConnectionSetup.java
package org.elkoserver.foundation.server;

import org.elkoserver.foundation.net.NetAddr;
import org.elkoserver.foundation.properties.ElkoProperties;
import org.elkoserver.foundation.server.metadata.AuthDesc;
import org.elkoserver.util.trace.Trace;
import org.elkoserver.util.trace.TraceFactory;

import java.io.IOException;

abstract class BaseConnectionSetup implements ConnectionSetup {
    final String host;
    private final AuthDesc auth;
    final boolean secure;
    final String propRoot;
    private final Trace trServer;
    private final Trace tr;
    protected TraceFactory traceFactory;
    final String bind;
    final Trace msgTrace;

    BaseConnectionSetup(String label, String host, AuthDesc auth, boolean secure, ElkoProperties props, String propRoot, Trace trServer, Trace tr, TraceFactory traceFactory) {
        this.host = host;
        this.auth = auth;
        this.secure = secure;
        this.propRoot = propRoot;
        this.trServer = trServer;
        this.tr = tr;
        this.traceFactory = traceFactory;

        bind = props.getProperty(propRoot + ".bind", host);

        if (label != null) {
            msgTrace = traceFactory.comm.subTrace(label);
        } else {
            msgTrace = traceFactory.comm.subTrace("cli");
        }
    }

    @Override
    public NetAddr startListener() {
        NetAddr result;

        try {
            result = tryToStartListener();

            trServer.noticei("listening for " + getProtocol() +
                    " connections" + getConnectionsSuffixForNotice() + " on " +
                    getListenAddressDescription() +
                    (!bind.equals(getValueToCompareWithBind()) ? (" (" + bind + ")") : "") +
                    " (" + auth.mode() + ")" +
                    (secure ? " (secure mode)" : ""));
        } catch (IOException e) {
            tr.errorm("unable to open " + getProtocol() + getProtocolSuffixForErrorMessage() +
                    " listener " + propRoot +
                    " on requested host: " + e);
            throw new IllegalStateException();
        }

        return result;
    }

    abstract NetAddr tryToStartListener() throws IOException;

    abstract String getListenAddressDescription();

    abstract String getValueToCompareWithBind();

    String getConnectionsSuffixForNotice() {
        return "";
    }

    String getProtocolSuffixForErrorMessage() {
        return "";
    }
}

File: ServerCore/src/main/kotlin/org/elkoserver/foundation/server/BaseConnectionSetup.kt
package org.elkoserver.foundation.server

import org.elkoserver.foundation.net.NetAddr
import org.elkoserver.foundation.properties.ElkoProperties
import org.elkoserver.foundation.server.metadata.AuthDesc
import org.elkoserver.util.trace.Trace
import org.elkoserver.util.trace.TraceFactory
import java.io.IOException

internal abstract class BaseConnectionSetup(label: String?, val host: String, private val auth: AuthDesc, val secure: Boolean, props: ElkoProperties, val propRoot: String, private val trServer: Trace, private val tr: Trace, protected var traceFactory: TraceFactory) : ConnectionSetup {
    val bind: String = props.getProperty("$propRoot.bind", host)

    var msgTrace = if (label != null) {
        traceFactory.comm.subTrace(label)
    } else {
        traceFactory.comm.subTrace("cli")
    }

    override fun startListener(): NetAddr {
        val result: NetAddr
        try {
            result = tryToStartListener()
            trServer.noticei("listening for $protocol connections$connectionsSuffixForNotice on $listenAddressDescription${(if (bind != valueToCompareWithBind) " ($bind)" else "")} (${auth.mode})${if (secure) " (secure mode)" else ""}")
        } catch (e: IOException) {
            tr.errorm("unable to open $protocol$protocolSuffixForErrorMessage listener $propRoot on requested host: $e")
            throw IllegalStateException()
        }
        return result
    }

    @Throws(IOException::class)
    abstract fun tryToStartListener(): NetAddr
    abstract val listenAddressDescription: String
    abstract val valueToCompareWithBind: String
    open val connectionsSuffixForNotice: String
        get() = ""

    open val protocolSuffixForErrorMessage: String
        get() = ""

}
